db/src/upsert_resolution.rs

```diff
@@ -241,7 +271,44 @@ impl Generation {
             }
         }
 
-        temp_ids
+        // Now we union-find all the known tempids.  Two tempids are unioned if they both appear as
+        // the entity of an `[a v]` upsert, including when the value column `v` is itself a tempid.
+        let mut uf = unionfind::UnionFind::new(temp_ids.len());
+
+        // The union-find implementation from petgraph operates on contiguous indices, so we need to
+        // maintain the map from our tempids to indices ourselves.
+        let temp_ids: BTreeMap<TempIdHandle, usize> = temp_ids.into_iter().enumerate().map(|(i, tempid)| (tempid, i)).collect();
+
+        debug!("need to label tempids aggregated using tempid_avs {:?}", tempid_avs);
+
+        for vs in tempid_avs.values() {
+            vs.first().and_then(|first| temp_ids.get(first)).map(|&first_index| {
+                for tempid in vs {
+                    temp_ids.get(tempid).map(|&i| uf.union(first_index, i));
+                }
+            });
+        }
+
+        debug!("union-find aggregation {:?}", uf.clone().into_labeling());
+
+        // Now that we have aggregated tempids, we need to label them using the smallest number of
+        // contiguous labels possible.
+        let mut tempid_map: BTreeMap<TempIdHandle, usize> = BTreeMap::default();
+
+        let mut dense_labels: indexmap::IndexSet<usize> = indexmap::IndexSet::default();
+
+        // We want to produce results that are as deterministic as possible, so we allocate labels
+        // for tempids in sorted order.  This has the effect of making "a" allocate before "b",
+        // which is pleasant for testing.
+        for (tempid, tempid_index) in temp_ids {
+            let rep = uf.find_mut(tempid_index);
```

Comment:
Does this handle transitive equivalence? That is, if `"foo"` is equivalent to `"bar"` because of `:foo/x 123`, and `"bar"` is equivalent to `"baz"` because of `:foo/y 456`, do `foo`, `bar, and `baz` all end up in the same bucket here?

Reply:
Yes, but let me make sure there's a test of exactly this.  You want:
```
[[:db/add "foo" :foo/x 123]
 [:db/add "bar" :foo/y 456]
 {:db/id "baz" :foo/x 123 :foo/y 456}]
```
to unify all of "foo", "bar", "baz"?  I guess we should test with those datoms present and not present.  I've added a test that covers this.

Reply:
In this PR you're making it so that

```
[:db/add "foo" :foo/x 123]
[:db/add "bar" :foo/x 123]
```

correctly unify; the next step is to make sure that


```
[:db/add "foo" :foo/x 123]
[:db/add "bar" :foo/x 123]
[:db/add "bar" :foo/y 456]
[:db/add "baz" :foo/y 456]
```

works. The subsequent insert itself will fail if it's not possible for all three entities to be the same, I think.